How Does *Fix With Copilot* work? (#1)
sandersn/sandersn.github.io: how-does-fix-with-copilot-work.md

Copilot might be able to guess, but *Fix With Copilot* can't even see `connect`'s code because it's imported. There's no way for anybody to figure out that the arguments are swapped. So, with just the error message to go on, Copilot makes a very local change: it reasons that `0` is the closest `number` equivalent of `false`. That technically fixes the error but another immediately pops up, because it's not a correct fix.

![Invoking AI](images/ai-bad-call-3.png)

When I look up and include the code for `connect` in the hidden prompt, Copilot can figure out that the arguments are swapped.

![Invoking AI](images/ai-bad-call-2.png)

## But really, how does Fix With Copilot work?

The first thing *Fix With Copilot* does is register as a fix provider for every kind of file. Every single one. Then, for each document you have open, it requests errors and adds *Fix With Copilot* to the list of available fixes for each.
<!-- see inlineChatCommand.ts:registerInlineChatCommands
and inlineChatCodeActions.ts:QuickFixesProvider -->

> When you click on *Fix With Copilot*, it invokes Copilot's inline chat with a hidden prompt constructed from JSX components. JSX&mdash;but not React. Copilot has its own renderer that generates plain text from JSX components, so *Fix With Copilot* is structured as a React-like hierarchy of small components. In addition to looking familiar, one clever property on all these components is `priority`. The renderer can drop low-priority components if the prompt gets too big. Here's an example of what the code looks like. `diagnostics` is the list of errors to be fixed:

Reviewer: There's a bit of a jump here - you might want to talk about how we could just stitch together raw text by writing our prompts in strings, JSX gives little bit of visual ergonomics and abstraction over priority.

On that note, priority doesn't come up again after you mention it - what in the below example is deprioritized?

Author: Hmph. Priorities are only specified for major blocks, like the whole block of diagnostic generation that I'm talking about. Maybe I'll omit that and instead explain the code a little bit.

<!-- see inlineChatFix4Prompt.tsx (old version for now) -->

```tsx
diagnostics.map((d, idx) => {
    const cookbook = getFixCookbook(documentContext.language.languageId, d);
    return <>
        <DiagnosticDescription 
          diagnostic={d} 
          cookbook={cookbook} 
          showLineNumbers={this.props.showLineNumbers ?? false} 
          maxLength={LINE_CONTEXT_MAX_SIZE} 
          documentContext={documentContext} />
        <DiagnosticRelatedInfo diagnostic={d} cookbook={cookbook} document={documentContext.document} />
        <DiagnosticSuggestedFix cookbook={cookbook} />
    </>;
})
```

What components are there? The main ones are:

- Include code surrounding the error.
- Format the error nicely.
- Provide instructions to Copilot.

> By default *Fix With Copilot* includes the entire file around the error. Of course, most of this time the entire file won't fit in the prompt. Instead, it includes the code in current function and summarizes code outside. The summary is constructed by X.

Reviewer: X?

Author: Uh, that's where I quit writing when I pushed the commit.

I realise now that's ambiguous with the social network formerly known as Twitter.

So next it tries to summarise the code outside the current function by selectively deleting code. By default it deletes all function bodies outside the current one.
(By default it divides the file into blocks and builds a tree out of it. Then each block gets assigned a weight based on distance from the error, and nesting depth. Blocks below a cutoff are omitted, so the prompt will include code near the error and less and less detail further away. 

    TODO: Find out from Martin how commonly this method is the one used, probably only describe the simple or complex one. This would be a good place for a diagram to explain the complex method.

> Formatting the error starts with the error text itself, unless my cookbook offers a replacement. But it also includes a couple of sub-components. The first includes code from related error spans. Language services can attach related information to an error, and Copilot will include the function that contains each related location. The other subcomponent is example usages, which is what I showed earlier for fixing Typescript errors. TODO: Make this longer, better structured and more readable.

Reviewer: The cookbook is implicitly mentioned in the code above, but this is the first mention otherwise.

Reviewer: ```suggestion
Formatting the error starts with the error text itself (unless our cookbook offers a replacement). But it also includes a couple of sub-components. The first includes code from related error spans. Language services can attach related information to an error, and Copilot will include the function that contains each related location. The other subcomponent is example usages, which is what I showed earlier for fixing TypeScript errors. TODO: Make this longer, better structured and more readable.
```

Author: Yeah, I decided to go with 'nudge' and 'instructions' rather than cookbook. I'm not sure it's the right choice but cookbook doesn't seem quite accurate to me either.

Once every block has a distance and a depth, each one gets a score and everything above a certain score gets abbreviated. To keep the example simple, let's just add `distance` and `depth` to get the score. Then you can see that anything with a scor above 3 is abbreviated. The body of `while (node)` was `distance:1 + depth:3` and the body of `interface Node<T>` was `distance:2 + depth:2`, so both get abbreviated.

### Format Error Message

Formatting the error starts with the error text itself, then adds a couple of sub-components: `DiagnosticRelatedInfo` and `DiagnosticSuggestedFix`.

> `DiagnosticRelatedInfo` itself adds two kinds of information: the first is any related code that the error itself contains. For example, the TypeScript error  *"Duplicate declaration"* includes the duplicated declaration as related info. The second kind of information is the source of imported functions in calls with bad argument types. *Fix With Copilot* finds the source of a called function in a language-independent way, using VS Code's language service API, although right now the code only applies to TypeScript.

Author: This paragraph should expand on how source is retrieved.  If the code is small enough, include a code sample of tree sitter and language service usage.

# How Does *Fix With Copilot* Work?
In VS Code, Copilot has a feature called *Fix With Copilot* that offers to fix any error in the editor with AI. It's like a code fix for every error imaginable, with the caveat that Copilot is just trying its best.

> 	Show an example here.

Reviewer: ```suggestion
If you've used GitHub Copilot, you might have noticed it offers a new option in the quick fix menu: *Fix with Copilot*.

	Show an example here.

*Fix With Copilot* offers to fix any error in the editor with AI. It's like having a quick fix for every error imaginable (with the caveat that Copilot is just trying its best - and aren't we all?).
```

> The Copilot extension intercepts every single error (diagnostic) that vscode gets from a language service

Reviewer: ```suggestion
The Copilot extension intercepts every single error (or "diagnostic") that vscode gets from a language service
```

Author: Reworded to avoid 'diagnostic' in the text; I'll need to define it if I leave in the JSX code snippet in the implementation section.

(TODO: Omit or find out whether it's actually subscribed to every one or if there is a generic subscription).

> Then it builds a prompt that, at its core, says "There is an error with ***Error text here*** in this code ***Code here***. Please fix it."

Reviewer: ```suggestion
Then it builds a prompt that, at its core, says "There is an error with ***\[\[error text here]]*** in this code ***\[\[code here]]***. Please fix it."
```

This works well for many errors, because the error implies a fix, and the fix is simple enough for Copilot to figure out. 

## Targeted Fixes For Specific Errors

> Fix With Copilot fails when the context or the error are misleading. Most errors in Typescript and ESLint, which I work with day-to-day, actually try to avoid prescribing a fix in cases where it's not obvious. That's the opposite of what Copilot needs -- if the error isn't helpful, then all it has to use is the context of surrounding code. For example, ESLint has a great rule that detects duplicate if/else predicates:

Reviewer: > Fix With Copilot fails when the context or the error are misleading.

You might say this is a problem for asking AI for help in general.

Also, I don't know if it's that the context or error are misleading. More that maybe they're vague about why the error exists, or are non-prescriptive about what should be done.

Author: +1 on AI in general.
There are cases when the errors are misleading vs non-prescriptive. I'll mention that distinction. Generally, I need to tease apart the discussion of code-context vs error here.

![Invoking AI](images/ai-no-dupe-if-1.png)

> The error message suggests that the duplicate is extraneous. Which it is, from a strictly control-flow perspective. But if we stop and look at *intent*, in my experience a duplicated `if/else` or `switch case` is the result of copy/paste/modifying cases until you get enough--and then forgetting to change the last one. That's why this ESLint rule is so great, because it catches attentional mistakes. Still, the fix is to make the duplicate case *different* from the others, not to delete.

Reviewer: It's hard to imagine what you're trying to showcase since these images weren't included in the PR. Still, I'd play with the ordering here:

```suggestion
The error message suggests that the duplicate `if` check is extraneous. That's technically true... but what happens if we ask Copilot for a fix here?

[!Copilot suggesting deleting the `if` check entirely](...)

Copilot manages to get rid of the error -- *by deleting the entire block*. But if that was the right solution, we'd be better off avoiding coding errors by not writing any code (and just to check in, have you gone outside today?).

To get the best fix, we need to consider the intent -- *why* might the code be duplicated? How did the author get into this situation? And a little bit of empathy goes a long way here. I know when I've written a series of `if/else/if` checks, or `case`s in a `switch`, I tend to just copy/paste and modify the blocks -- but I also tend to forget to update that code with a new value to check against. And that's actually why this ESLint rule is so great: because it catches attentional mistakes.

So the best fix isn't to delete the code, it's to make the duplicate case *different* from the others.
```

Author: The rich diff should show the images, since they're already in the repo.

I re-ordered the presentation according to your high-level suggestion -- but didn't expand the text as much as this suggestion. Let me know if you think it needs more explanation here.
